src/core/meshcop/border_agent.cpp

```diff
@@ -0,0 +1,606 @@
+
+    SuccessOrExit(error = CopyPayload(aMessage, message));
+
+    messageInfo.SetPeerAddr(mCommissionerAddr);
+    messageInfo.SetPeerPort(mCommissionerUdpPort);
+
+    SuccessOrExit(error = mNetif.GetBASecureCoapServer().SendMessage(*message, messageInfo));
+
+    otLogCritMeshCoP(GetInstance(), "sent MGMT_COMMISSIONER_SET response to commissioner");
+
+exit:
+    (void) aMessageInfo;
+    if (error != kThreadError_None && message != NULL)
+    {
+        message->Free();
+    }
+    otLogFuncExit();
+}
+
+void BorderAgent::HandleRelayReceive(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                                     const otMessageInfo *aMessageInfo)
+{
+    static_cast<BorderAgent *>(aContext)->HandleRelayReceive(
+        *static_cast<Coap::Header *>(aHeader), *static_cast<Message *>(aMessage),
+        *static_cast<const Ip6::MessageInfo *>(aMessageInfo));
+}
+
+void BorderAgent::HandleRelayReceive(Coap::Header &aHeader, Message &aMessage,
+                                     const Ip6::MessageInfo &aMessageInfo)
+{
+    ThreadError error = kThreadError_None;
+    Coap::Header header;
+    Message *message = NULL;
+    Ip6::MessageInfo messageInfo;
+
+    otLogFuncEntry();
+    otLogCritMeshCoP(GetInstance(), "received Relay Recevie from Joiner Router");
+
+    header.Init(kCoapTypeConfirmable, kCoapRequestPost);
+    header.SetToken(aHeader.GetToken(), aHeader.GetTokenLength());
+    header.AppendUriPathOptions(OPENTHREAD_URI_RELAY_RX);
+    header.SetPayloadMarker();
+
+    VerifyOrExit((message = mNetif.GetBASecureCoapServer().NewMeshCoPMessage(header)) != NULL, error = kThreadError_NoBufs);
+
+    SuccessOrExit(error = CopyPayload(aMessage, message));
+
+    messageInfo.SetPeerAddr(mCommissionerAddr);
+
+    otLogCritMeshCoP(GetInstance(), "send Relay Receive to Commissioner");
+
+    SendCommissionerRelayReceive(*message, messageInfo);
+
+
+exit:
+    (void)aMessageInfo;
+    if (error != kThreadError_None && message != NULL)
+    {
+        message->Free();
+    }
+    otLogFuncExit();
+}
+
+
+ThreadError BorderAgent::SendCommissionerRelayReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    ThreadError error = kThreadError_None;
+
+    otLogFuncEntry();
+
+    SuccessOrExit(error = mNetif.GetBASecureCoapServer().SendMessage(aMessage, aMessageInfo));
+
+    otLogCritMeshCoP(GetInstance(), "sent Realy Receive to Commissioner");
+
+exit:
+    (void) aMessageInfo;
+    otLogFuncExitErr(error);
+    return error;
+}
+
+void BorderAgent::HandleTimer(void *aContext)
+{
+    static_cast<BorderAgent *>(aContext)->HandleTimer();
+}
+
+void BorderAgent::HandleTimer(void)
+{
+    ThreadError error = kThreadError_Error;
+    PSKcTlv *pskc;
+    otLogFuncEntry();
+
+    pskc = static_cast<PSKcTlv *>(mNetif.GetActiveDataset().GetNetwork().Get(Tlv::kPSKc));
+    if (pskc != NULL)
+    {
+        SuccessOrExit(error = mNetif.GetBASecureCoapServer().SetPsk(reinterpret_cast<const uint8_t *>(pskc->GetPSKc()),
```

Comment:
To make this more readable, can you do the function call and assignment on one line, and then the error check on another?